cpu/rpx0xx/periph/usbdev.c

```diff
@@ -0,0 +1,746 @@
+static inline bool _reg_get_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    uint32_t val = *reg;
+    return val & 1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1));
+}
+
+/* Clear the ep-related bit from a status register. */
+static inline void _reg_clear_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    io_reg_atomic_clear(reg, 
+                        1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1)));
+}
+
+static inline __IOM uint32_t * _get_ep_buf_ctrl_reg(uint8_t ep_num,
+                                                    usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_buf_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_buf_ctrl_reg(ep_num);
+    }
+}
+
+static inline __IOM uint32_t * _get_ep_ctrl_reg(uint8_t ep_num,
+                                                usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_ctrl_reg(ep_num);
+    }
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir,
+                                      uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_buf_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir, 
+                                          uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_buf_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+}
+
+/* Get DPRAM buffer offset. */
+static inline uintptr_t _dpram_offset(uint8_t *buf)
+{
+    return (uintptr_t) buf ^ (uintptr_t) USBCTRL_DPRAM_BASE;
+}
+
+static rpx0xx_usb_ep_t *_get_ep(uint8_t ep_num, usb_ep_dir_t dir)
+{
+    return &_hw_usb_dev.hw_ep[2 * ep_num + (dir == USB_EP_DIR_OUT ? 0 : 1)];
+}
+
+static void _poweron(void)
+{
+    periph_reset(RESETS_RESET_usbctrl_Msk);
+    periph_reset_done(RESETS_RESET_usbctrl_Msk);
+    io_reg_atomic_set(&USBCTRL_REGS->MAIN_CTRL,
+                      USBCTRL_REGS_MAIN_CTRL_CONTROLLER_EN_Msk);
+    io_reg_atomic_set(&USBCTRL_REGS->USB_MUXING,
+                      USBCTRL_REGS_USB_MUXING_TO_PHY_Msk);
+}
+
+/*
+static void _poweroff(void)
+{
+    io_reg_atomic_set(&USBCTRL_REGS->MAIN_CTRL, 0);
+    periph_reset(RESETS_RESET_usbctrl_Msk);
+}
+*/
+
+void usbdev_init_lowlevel(void)
+{
+    DEBUG("[rpx0xx usb] Call usbdev_init_lowlevel()\n");
+    memset(&_hw_usb_dev, 0, sizeof(rpx0xx_usb_dev_t));
+    _hw_usb_dev.dev.driver = &driver;
+}
+
+usbdev_t *usbdev_get_ctx(unsigned num)
+{
+    (void)num;
+    return &_hw_usb_dev.dev;
+}
+
+static void _enable_irq(void) {
+    io_reg_atomic_set(&USBCTRL_REGS->INTE,
+                      USBCTRL_REGS_INTE_BUS_RESET_Msk |
+                      USBCTRL_REGS_INTE_DEV_SUSPEND_Msk |
+                      USBCTRL_REGS_INTE_DEV_CONN_DIS_Msk);
```

Comment:
You should probably add `USBCTRL_REGS_INTE_BUFF_STATUS_Msk` and `USBCTRL_REGS_INTE_SETUP_REQ_Msk`
too